File: app/src/main/java/org/miaowo/miaowo/fragment/setting/AppSetting.java
package org.miaowo.miaowo.fragment.setting;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Switch;
import android.widget.TextView;

import com.sdsmdg.tastytoast.TastyToast;

import org.json.JSONArray;
import org.json.JSONObject;
import org.miaowo.miaowo.R;
import org.miaowo.miaowo.api.API;
import org.miaowo.miaowo.bean.data.User;
import org.miaowo.miaowo.root.BaseActivity;
import org.miaowo.miaowo.root.BaseFragment;
import org.miaowo.miaowo.util.LogUtil;
import org.miaowo.miaowo.util.SpUtil;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;

/**
 * 设置-用户设置
 * Created by luqin on 17-5-11.
 */

public class AppSetting extends BaseFragment {
    public static final String SETTING_APP_TAB = "square_tab_mode";
    public static final String SETTING_APP_CLEAN = "auto_clean_token";

    @BindView(R.id.sw_clean) Switch swClean;
    @BindView(R.id.sw_tab) Switch swTab;
    @BindView(R.id.rv_tokens) RecyclerView rvTokens;

    private API mApi;
    private User mUser;
    private String mToken;

    public AppSetting() {
    }

    public static AppSetting newInstance() {
        Bundle args = new Bundle();
        AppSetting fragment = new AppSetting();
        fragment.setArguments(args);
        return fragment;
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_setting_app, container, false);
    }

    @Override
    public void initView(View view) {
        mApi = new API();
        mUser = API.loginUser;
        mToken = API.token;
        SpUtil mSp = SpUtil.defaultSp();

        swTab.setChecked(mSp.getBoolean(SETTING_APP_TAB, false));
        swClean.setChecked(mSp.getBoolean(SETTING_APP_CLEAN, true));
        swTab.setOnClickListener(v -> mSp.putBoolean(SETTING_APP_TAB, swTab.isChecked()));
        swClean.setOnClickListener(v -> mSp.putBoolean(SETTING_APP_CLEAN, swClean.isChecked()));

        rvTokens.setLayoutManager(new LinearLayoutManager(getContext()) {
            @Override
            public boolean canScrollVertically() {
                return false;
            }
        });
        TokenAdapter adapter = new TokenAdapter();
        rvTokens.setAdapter(adapter);

        mApi.useAPI(API.APIType.USERS, mUser.getUid() + "/tokens", API.Method.GET, true, null, (call, response) -> {
            try {
                JSONObject object = new JSONObject(response.body().string());
                if (!"ok".equals(object.getString("code"))) throw new Exception("获取失败");
                else {
                    JSONArray tokenArray = object.getJSONObject("payload").getJSONArray("tokens");
                    ArrayList<String> tokens = new ArrayList<>();
                    for (int i = 0; i < tokenArray.length(); i++) tokens.add(tokenArray.getString(i));
                    adapter.update(tokens);
                }
            } catch (Exception e) {
                BaseActivity.get.toast(e.getMessage(), TastyToast.ERROR);
            }
        });
    }

    private class TokenAdapter extends RecyclerView.Adapter {
        List<String> mTokens = new ArrayList<>();

        @Override
        public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            return new RecyclerView.ViewHolder(LayoutInflater.from(BaseActivity.get).inflate(R.layout.list_token, parent, false)) {};
        }

        @Override
        public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
            View v = holder.itemView;
            TextView tv_token = (TextView) v.findViewById(R.id.token);
            String token = mTokens.get(position);
            tv_token.setText(token);
            boolean equals = token.equals(mToken);
            v.findViewById(R.id.thisToken).setVisibility(equals ? View.VISIBLE : View.GONE);
            v.findViewById(R.id.remove).setVisibility(equals ? View.GONE : View.VISIBLE);
            v.findViewById(R.id.remove).setOnClickListener(v1 -> {
                if (equals) v1.setVisibility(View.GONE);
                else mApi.useAPI(API.APIType.USERS, mUser.getUid() + "/tokens/" + token, API.Method.DELETE, true, null, (call, response) -> {
                    try {
                        JSONObject object = new JSONObject(response.body().string());
                        if ("ok".equals(object.getString("code"))) {
                            mTokens.remove(position);
                            BaseActivity.get.runOnUiThreadIgnoreError(this::notifyDataSetChanged);
                        }
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                });
            });
        }

        @Override
        public int getItemCount() {
            return mTokens.size();
        }

        public void update(List<String> tokens) {
            mTokens = tokens;
            BaseActivity.get.runOnUiThreadIgnoreError(this::notifyDataSetChanged);
        }
    }
}

File: app/src/main/java/org/miaowo/miaowo/fragment/setting/AppSetting.kt
package org.miaowo.miaowo.fragment.setting

import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.blankj.utilcode.util.ActivityUtils.getTopActivity
import kotlinx.android.synthetic.main.fragment_setting_app.*
import okhttp3.ResponseBody
import org.json.JSONObject
import org.miaowo.miaowo.API
import org.miaowo.miaowo.App
import org.miaowo.miaowo.R
import org.miaowo.miaowo.base.ListHolder
import org.miaowo.miaowo.base.extra.error
import org.miaowo.miaowo.base.extra.inflateId
import org.miaowo.miaowo.other.ActivityCallback
import org.miaowo.miaowo.other.Const
import org.miaowo.miaowo.other.template.EmptyCallback
import retrofit2.Call
import retrofit2.Response

/**
 * 设置-用户设置
 * Created by luqin on 17-5-11.
 */

class AppSetting : Fragment() {


    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflateId(R.layout.fragment_setting_app, inflater, container)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        sw_clean.isChecked = App.SP.getBoolean(Const.SP_CLEAN_TOKENS, true)
        sw_clean.setOnClickListener { App.SP.put(Const.SP_CLEAN_TOKENS, sw_clean.isChecked) }

        sw_hide.isChecked = App.SP.getBoolean(Const.SP_HIDE_BODY, false)
        sw_hide.setOnClickListener {
            val ret = sw_hide.isChecked
            App.SP.put(Const.SP_HIDE_BODY, ret)
            val visibility = if (ret) View.GONE else View.VISIBLE
            title_hide.visibility = visibility
            rg_hide.visibility = visibility
        }

        rg_hide.check(when (App.SP.getInt(Const.SP_SHOW_TYPE, Const.CBODY_LAST)) {
            Const.CBODY_FIRST -> R.id.rb_first
            Const.CBODY_CONTENT -> R.id.rb_content
            else -> R.id.rb_last
        })
        rg_hide.setOnCheckedChangeListener { _, checkedId ->
            App.SP.put(Const.SP_SHOW_TYPE, when (checkedId) {
                R.id.rb_first -> Const.CBODY_FIRST
                R.id.rb_content -> Const.CBODY_CONTENT
                else -> Const.CBODY_LAST
            })
        }

        rv_tokens.layoutManager = object : LinearLayoutManager(context) {
            override fun canScrollVertically() = false
        }
        val adapter = TokenAdapter()
        rv_tokens.adapter = adapter
    }

    private class TokenAdapter : RecyclerView.Adapter<ListHolder>() {

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) =
                ListHolder(R.layout.list_token, parent)

        override fun onBindViewHolder(holder: ListHolder, position: Int) {
            val tk = API.token[position]
            holder.find<TextView>(R.id.token)?.text = tk
            holder[R.id.remove]?.setOnClickListener {
                API.Users.tokenRemove(tk, API.user.uid).enqueue(object : ActivityCallback<ResponseBody>(getTopActivity()) {
                    override fun onResponse(call: Call<ResponseBody>?, response: Response<ResponseBody>?) {
                        val obj = JSONObject(response?.body()?.string())
                        if (obj.getString("code").toUpperCase() == Const.RET_OK) {
                            var err = obj.getString("message")
                            if (err.isNullOrBlank()) err = obj.getString("code")
                            if (err.isNullOrBlank()) err = activity.getString(R.string.err_ill)
                            getTopActivity().error(err)
                        } else {
                            API.token.remove(tk)
                            notifyDataSetChanged()
                        }
                    }
                })
            }
        }

        override fun getItemCount() = API.token.size
    }

    companion object {
        fun newInstance(): AppSetting {
            val args = Bundle()
            val fragment = AppSetting()
            args.putString(Const.TAG, fragment.javaClass.name)
            fragment.arguments = args
            return fragment
        }
    }
}
